### ISSUE TITLE
`trio.run` takes a long time to start

### ISSUE THREAD

**Issue by SysuJayce** [python-trio/trio#2770]
Hello there,
  I start millions http request with trio's `nursery.start_soon`, but it seems the program gets stuck for several minutes after I run `trio.run`.
  I wonder what happens after I run `trio.run`. How can I make the program run smoothly?

**Comment by TeamSpen210** [comment]
Can you show your code? Trio should be able to handle lots of tasks, maybe not millions. But if they're all HTTP requests, them all trying to actually run simultaneously would cause a lot of contention and whatnot. Are you using a `CapacityLimiter`?

**Comment by SysuJayce** [comment]
> 

Yes I use a `CapacityLimiter ` because the server has a qps limit

**Comment by A5rocks** [comment]
Sorry, I don't quite understand: do you run a ton of tasks and *then* trio.run? Or do you run trio.run and then after several minutes of a ton of spawning tasks the program hangs?

**Comment by SysuJayce** [comment]
> Sorry, I don't quite understand: do you run a ton of tasks and _then_ trio.run? Or do you run trio.run and then after several minutes of a ton of spawning tasks the program hangs?

I use 2 `CapacityLimiter` to control 2 different http request's QPS.
The dataframe has 10m rows, which means I will send 20m http requests totally.

The exact behavior is that when `trio.run` executes, the program hangs for several minutes before the http requests send to the server.

Is this behavior expected? I expect that within the `CapacityLimiter` context, the task gets executed immediately after `nursery.start_soon`

Here is the snippet.
```
async def worker(limit=100):
    limiter1 = trio.CapacityLimiter(limit)
    limiter2 = trio.CapacityLimiter(limit)
    async with httpx.AsyncClient() as client:
        async with trio.open_nursery() as nursery:
            for idx in df.index:  # df is a pandas' DataFrame
                nursery.start_soon(query, idx, client, limiter1, url1)
                nursery.start_soon(query, idx, client, limiter2, url2)

trio.run(worker)
```

**Comment by A5rocks** [comment]
Oh I think I see what's going on: trio can only actually execute `query`s once it hits an `await`... which you never do until you've spawned *all* the tasks!

You could stick an `await trio.lowlevel.checkpoint()` (or `await trio.sleep(0)`) I think? Or, well, I forgot the exact interface for `CapacityLimiter`s but you might be able to wait for them outside of the `nursery.start_soon` thing?

**Comment by SysuJayce** [comment]
> Oh I think I see what's going on: trio can only actually execute `query`s once it hits an `await`... which you never do until you've spawned _all_ the tasks!
> 
> You could stick an `await trio.lowlevel.checkpoint()` (or `await trio.sleep(0)`) I think? Or, well, I forgot the exact interface for `CapacityLimiter`s but you might be able to wait for them outside of the `nursery.start_soon` thing?

@A5rocks Thanks for the info. I'm not familiar to async program. Can you point out where should I insert the `await trio.sleep(0)`?

BTW, the `query` takes about 5-10ms usually.
```
async def worker(limit=100):
    limiter1 = trio.CapacityLimiter(limit)
    limiter2 = trio.CapacityLimiter(limit)
    async with httpx.AsyncClient() as client:
        async with trio.open_nursery() as nursery:
            for idx in df.index:  # df is a pandas' DataFrame
                nursery.start_soon(query, idx, client, limiter1, url1)
                nursery.start_soon(query, idx, client, limiter2, url2)
            await trio.sleep(0)  # I'm not sure where to insert this code
trio.run(worker)
```

**Comment by A5rocks** [comment]
Errr, basically, `await` "yields" back to the event loop: the *only* time trio can do some stuff (e.g. scheduling some stuff, running another task, switching to `query`, ...) is at an `await`.

You should put it right after both `nursery.start_soon`s, at the same indentation level.

**Comment by A5rocks** [comment]
Here's a much longer explanation partially about how trio can schedule stuff only at `await`s: https://trio.readthedocs.io/en/stable/design.html#cancel-points-and-schedule-points

**Comment by SysuJayce** [comment]
> 



> Errr, basically, `await` "yields" back to the event loop: the _only_ time trio can do some stuff (e.g. scheduling some stuff, running another task, switching to `query`, ...) is at an `await`.
> 
> You should put it right after both `nursery.start_soon`s, at the same indentation level.

Good thing: I add `await trio.sleep(0)` after `nursery.start_soon` and the `query` executes immediately.  
Bad thing: the total run time become a little longer. 

Maybe `await trio.sleep(0)` slow down the program?
 Or should we just wait for all the tasks spawned?

**Comment by A5rocks** [comment]
`trio.sleep` is somewhat slow, yeah. It's all the work around tossing tasks around that trio has to do!

However, given your workload, I would be cautious about waiting for all tasks to be spawned: that takes minutes as you've pointed out, which are minutes you could be using given your qps limit.

Instead, maybe you could checkpoint only every so often? Maybe every 100 iterations?

**Comment by smurfix** [comment]
The problem is that given your workload what you're doing is, umm, far from optimal. You're starting a gazillion tasks, just to let all of them wait on the same limiter. Setting up all these tasks is a whole lot of unnecessary work.

A better idea in your case is to open an in-memory object stream and a couple of workers. Start as many workers as you have capacity for (`2*limit` in this case), no need for explicit limiters. Let the workers iterate over the stream. Close the send side of the stream after your loop so the workers know to stop.

**Comment by SysuJayce** [comment]
Thank you!
I have tried to call `await trio.lowlevel.checkpoint()` every 100 iters and it is a little faster than `trio.sleep` every iter.
But the total run time is still not improved. @A5rocks 

I ask chatgpt to refactor the snippet according to advice from @smurfix .
And it gives me an example to use `trio.open_memory_channel`

**Comment by smurfix** [comment]
> I ask chatgpt to refactor the snippet according to advice from @smurfix .

This code still starts one worker per data item. It also gets the call to `open_memory_channel` wrong.
Sorry for being harsh here, but ChatGPT is not a replacement for understanding what you're doing.

Translating my remarks into Python:

```python3
async def worker(chan, client):
    async for idx in chan:
        # no need to further parallelize these
        await do_a_query(idx, client, url1)
        await do_a_query(idx, client, url2)

async def main(limit=100):
    async with httpx.AsyncClient(limits= httpx.Limits(max_connections=limit)) as client, trio.open_nursery() as nursery:       
        send_channel, receive_channel = trio.open_memory_channel(limit)
        for _ in range(min(limit,len(df.index))):
            nursery.start_soon(worker, receive_channel, client)
        with send_channel:
            for idx in df.index:  # df is a pandas' DataFrame
                await send_channel.send(idx)
                
trio.run(main)
```

**Comment by SysuJayce** [comment]
> > I ask chatgpt to refactor the snippet according to advice from @smurfix .
> 
> The code from GPT still starts one worker per data item. It also gets the call to `open_memory_channel` wrong. Sorry for sounding somewhat harsh here, but ChatGPT is not a replacement for understanding what you're doing.
> 
> Translating my remarks into Python (untested, obviously):
> 
> ```python
> async def worker(chan, client):
>     async for idx in chan:
>         # no need to further parallelize these
>         await do_a_query(idx, client, url1)
>         await do_a_query(idx, client, url2)
> 
> async def main(limit=100):
>     async with httpx.AsyncClient(limits= httpx.Limits(max_connections=limit)) as client, trio.open_nursery() as nursery:       
>         send_channel, receive_channel = trio.open_memory_channel(limit)
>         for _ in range(min(limit,len(df.index))):
>             nursery.start_soon(worker, receive_channel, client)
>         with send_channel:
>             for idx in df.index:  # df is a pandas' DataFrame
>                 await send_channel.send(idx)
>                 
> trio.run(main)
> ```

Thank you very much @smurfix 
It really helps a lot. The time cost reduced ~15%.

Btw, I also realize that ChatGPT sometimes gives me wrong answer lol.
